scipy/optimize/minpack.py

```diff
@@ -663,10 +673,25 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         else:
             xdata = np.asarray(xdata)
 
-    weights = 1.0 / asarray(sigma) if sigma is not None else None
-    func = _wrap_func(f, xdata, ydata, weights)
+    # If usecov is not None, redefine the problem
```

Comment:
I suggest the following sequence for `sigma` verification and processing:

1. If `not None`, wrap it to ndarray: `sigma = np.asarray(sigma)`.
2. Check that its shape is either `(M,)` or `(M, M)`. If not, raise `ValueError("sigma has incorrect shape.")` (or something like that).
3. Then assign to variable, say `T` (for " transform"), depending on the case:
   1. `sigma is None` -> `T = None`.
   2. `sigma is 1-d` -> `T = 1 / sigma`.
   3. `sigma is 2-d` -> `T = cholesky(sigma)`. Catch `LinalgError` from `cholesky` and report that "`sigma`     must be positive definite".

Then `_wrap_func` and `_wrap_jac` can accept a single parameter `T`, which will be handled accordingly.